### ISSUE TITLE
Distorted main screen if second screen is rotated

### ISSUE THREAD

**Issue by Henning0** [mate-desktop/marco#578]
Hello. I get a distorted main screen as soon as I set my second screen as rotated. This only happens with composite on. I've tested a few resolution/refresh rate combinations for both screens, but this doesn't seem to make a difference.

I've no clue if this is a useful information, but I gave Metacity a short try and it worked fine with composite on in this regard.

What really strange is, I had this setup running fine. The only difference is that I upgraded my GPU from a RX 580 8GB to a RX 5700 XT. The main screen is a Acer 2560x1440/144hz, the second screen is a Dell 1920x1200/60hz.

Since I can't say for sure if this is your bug and in case it isn't, I would really appreciate it if you could hint me to where the problem maybe originate from.

#### Expected behaviour
![fine](https://user-images.githubusercontent.com/28120455/72300312-b7b55d00-3663-11ea-8b35-2d09690c9e03.jpg)

#### Actual behaviour
![distorted](https://user-images.githubusercontent.com/28120455/72300291-a79d7d80-3663-11ea-8aa8-ce4c7e270c61.jpg)
The areas which look like corrupted memory flickers I guess every rendered frame. The black area on the left side shows distorted parts which normally belong on the second screen.

#### Steps to reproduce the behaviour
1. Use marco with composite on
2. Have a dual screen setup
3. Set up second screen as rotated to left or right

#### MATE general version
`MATE Desktop Environment 1.22.2`

#### Package version
`marco 1.22.3`

#### Linux Distribution
Manjaro Stable
`Linux 5.4.6-2-MANJARO`

#### Link to downstream report of your Distribution
None

**Comment by Jildridge** [comment]
Hello,
I am experiencing currently also problems with that kind of garbled screens únder mate.
Did you already try to switch the composit of, using just marco; so I did and the screen looks stable afterwards.

**Comment by Henning0** [comment]
Yes, with compositing off, it works without problems.

But I'd like to avoid that. Since the default theme of Manjaro relies on the shadow around windows to allow resizing them and there is screen tearing.

As this started for me with upgrading to an AMD Navi GPU, do you also happen to have an AMD Navi GPU in use?

Btw, for everyone else reading, I can still reproduce the problem with the latest versions from Manjaro:

`MATE Desktop Environment 1.24.0`
`marco 1.24.0`
`Linux 5.7.0-3-MANJARO`

**Comment by raveit65** [comment]
I never saw that with using nvidia driver.
Sorry, i don't have a AMD card for reproducing.

**Comment by jonathonf** [comment]
I can reliably replicate this with a laptop running with `amdgpu` and an external display and `marco` 1.24.1. Running with `metacity` instead works fine so this points to something with `marco`'s software compositing. Let me know what information I can provide to help debug this.

Lots of background, display layout combinations etc., here: https://forum.endeavouros.com/t/flickering-checker-board-using-amdgpu-laptop-with-external-display-and-marco/6966

**Comment by jonathonf** [comment]
I've managed to replicate this same symptom with Picom (from git https://github.com/yshui/picom/commit/9bb21fc726dc663c4e943e489148a6b343c79ee8) using `--experimental-backends` with `backend="xrender";` and setting `vsync=true;`. This combination can cause issues: https://github.com/yshui/picom/wiki/Vsync-Situation

That page indicates the issue might lie within having displays with different refresh rates, which does make sense - the display corruption does look similar to an "out of range" display. However, the displays are set to the same 60Hz rate within Display Settings so could this be due to a lower-level difference? e.g.:

```
$ xrandr --prop
...
   1920x1080     60.02 +
   1680x1050     60.02  
   1280x1024     60.02
...
   1680x1050     59.88*+  59.95  
   1280x1024     75.02    60.02  
   1440x900      59.90
...
```

This looks to be related to DRI/XPresent. Support for that was implemented a while ago: https://github.com/mate-desktop/marco/issues/326 , but then this probably explains why it's not seen on an NVIDIA card: https://github.com/mate-desktop/marco/pull/350#issuecomment-340037421

> nvidia proprietary drivers don't support DRI3.
No DRI3 --> no XPresent

**Comment by bjo81** [comment]
I have the same issue on a Thinkpad A275 with  amdgpu and had to disable xpresent usage on XFCE.

**Comment by Philippe734** [comment]
This issue is tracking on launchpad:
https://bugs.launchpad.net/ubuntu/+source/xserver-xorg-video-amdgpu/+bug/1873895

**Comment by jonathonf** [comment]
From that, it looks to have been fixed in `xf86-driver-amdgpu`:

https://gitlab.freedesktop.org/xorg/driver/xf86-video-amdgpu/-/issues/10

Now it's just a case of waiting (though I will be trying the patch from above against 19.1.0 😁).

**Comment by Philippe734** [comment]
> From that, it looks to have been fixed in `xf86-driver-amdgpu`:
> 
> https://gitlab.freedesktop.org/xorg/driver/xf86-video-amdgpu/-/issues/10
> 
> Now it's just a case of waiting (though I will be trying the patch from above against 19.1.0 grin).
> 
> Edit: Yup, that seems to fix it.

How to install or try the patch in Ubuntu?

**Comment by Philippe734** [comment]
Same issue with Ubuntu Mate HWE 20.04 with kernel 5.8 and latest AMD driver from oibaf PPA.

```
$ inxi -SMG
System:    Host: ACER-xk2308 Kernel: 5.8.0-38-generic x86_64 bits: 64 Desktop: MATE 1.24.0 
           Distro: Ubuntu 20.04.1 LTS (Focal Fossa) 
Machine:   Type: Desktop System: MSI product: MS-7798 v: 1.0 serial: <superuser/root required> 
           Mobo: MSI model: B75MA-P45 (MS-7798) v: 1.0 serial: <superuser/root required> BIOS: American Megatrends v: 1.9 
           date: 09/30/2013 
Graphics:  Device-1: Advanced Micro Devices [AMD/ATI] Navi 10 [Radeon RX 5600 OEM/5600 XT / 5700/5700 XT] driver: amdgpu 
           v: kernel 
           Display: x11 server: X.Org 1.20.9 driver: amdgpu,ati unloaded: fbdev,modesetting,radeon,vesa 
           resolution: 1366x768~60Hz 
           OpenGL: renderer: AMD Radeon RX 5600 XT (NAVI10 DRM 3.38.0 5.8.0-38-generic LLVM 11.0.1) 
           v: 4.6 Mesa 21.1.0-devel (git-7b48d5d 2021-01-16 focal-oibaf-ppa)
$ hwe-support-status --verbose
Your Hardware Enablement Stack (HWE) is supported until avril 2025
```

Please, let me know what to do, for testing what you want.

**Comment by Philippe734** [comment]
You may find the following information useful. It looks like Ubuntu needs to add these fixes ...
https://gitlab.freedesktop.org/xorg/driver/xf86-video-amdgpu/-/issues/10#note_769719

**Comment by jonathonf** [comment]
Yes - that's the issue I linked to earlier. This not a MATE bug, it's a bug in the amdgpu driver. This will be fixed upstream "at some point".

Until then you can request distro packages/maintainers to pick up the patch. I've tested it and it works.

**Comment by paulocoghi** [comment]
@jonathonf, could you explain how you applied the patch? Thanks!

_(Ubuntu Mate 20.04 here)_
